### ISSUE TITLE
Ansible is unable to connect in parallel to hosts in different groups behind different bastion hosts, fails randomly

### ISSUE THREAD

**Issue by plemelin** [ansible/ansible#79593]
### Summary

I'm using multiple inventories to connect to hosts at different locations. Even when gathering facts, the connection will randomly fails if more than one fork is used. 

The ssh config: 

```
Host A
  Hostname A.somedomain.com
  User user
  Port 12345
Host B
  Hostname B.somedomain.com
  User user
  Port 12345
```

The inventory files `inventories`:

A.yml
```
all:
  children:
    A:
      hosts:
        A_machine:
          ansible_host: machine
      vars:
        ansible_ssh_common_args: '-J A -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
        ansible_ssh_user: root
        ansible_ssh_password: abcdef
```

B.yml
```
all:
  children:
    B:
      hosts:
        B_machine:
          ansible_host: machine
      vars:
        ansible_ssh_common_args: '-J B -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
        ansible_ssh_user: root
        ansible_ssh_password: abcdef
```

The command that fails:

```
ANSIBLE_PIPELINING=1 ANSIBLE_HOST_KEY_CHECKING=False ansible-playbook -f 10 -i inventories -l '*_machine' playbooks/test_inventory.yml
```

Once verbose is enabled, the error shown will be: `Control master terminated unexpectedly`.

### Issue Type

Bug Report

### Component Name

ansible-core

### Ansible Version

```console
$ ansible --version
ansible [core 2.14.1]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/palemelin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  ansible collection location = /home/palemelin/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.8 (main, Nov  1 2022, 14:18:21) [GCC 12.2.0] (/usr/bin/python)
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
CONFIG_FILE() = /etc/ansible/ansible.cfg
```

### OS / Environment

`ansible-playbook` runs on archlinux

The bastion host can be `CentOS 7` or `Ubuntu 20.04`.

All `machine` hosts are archlinux hosts.

### Steps to Reproduce

With the aforementioned inventories, if I have 5 different environment, they start to fail randomly. 

### Expected Results

I expect to be able to run the tasks in parallel on the different `machine` hosts behind the different environment. 

### Actual Results

Sometimes connecting to A_machine will fail, sometimes it will be B_machine. 

I suspect it has to do with the fact that the ControlMaster is set to auto and than the `ansible_host` for `A_machine` and `B_machine` is set to `machine`?!

```console
<machine> ESTABLISH SSH CONNECTION FOR USER: root
<machine> SSH: EXEC sshpass -d25 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'User="root"' -o ConnectTimeout=10 -J A -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o 'ControlPath="/home/palemelin/.ansible/cp/d30ffb40ce"' machine '/bin/sh -c '"'"'/usr/bin/python3.10 && sleep 0'"'"''

fatal: [A_machine]: UNREACHABLE! => {
    "changed": false,
    "msg": "Data could not be sent to remote host \"machine\". Make sure this host can be reached over ssh: OpenSSH_9.1p1, OpenSSL 3.0.7 1 Nov 2022\r\ndebug1: Reading configuration data /home/palemelin/.ssh/config\r\ndebug1: /home/palemelin/.ssh/config line 3: Applying options for *\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug3: kex names ok: [diffie-hellman-group1-sha1]\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: Setting implicit ProxyCommand from ProxyJump: ssh -vvv -W '[%h]:%p' A\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 1209721\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 4\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Control master terminated unexpectedly\r\n",
    "unreachable": true
}
```
```


### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/devel/ISSUE_HELP.md)

**Comment by sivel** [comment]
This is likely related to, if not a duplicate of https://github.com/ansible/ansible/issues/76956

**Comment by plemelin** [comment]
> This is likely related to, if not a duplicate of, #76956

Well, I'm starting to doubt my googling skill for sure.

It is indeed a duplicate. 

Updating my inventory like so allowed me to work around the issue. 

```
all:
  children:
    A:
      hosts:
        A_machine:
          ansible_host: machine
          ansible_control_path: "/tmp/ansible-ssh-%%h-%%p-%%r-{{ inventory_hostname }}"
      vars:
        ansible_ssh_common_args: '-J A -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null'
        ansible_ssh_user: root
        ansible_ssh_password: abcdef
```

**Comment by bcoca** [comment]
closing as per above
